exercises/clock/src/clock.h

```diff
@@ -1,11 +1,19 @@
 #ifndef CLOCK_H
 #define CLOCK_H
 
+#include<stdbool.h>
+
 #define MAX_STR_LEN (5 + 1)     // "##:##\0"
 
-typedef char time_text_t[MAX_STR_LEN];
+typedef char clock_text_t[MAX_STR_LEN];
+typedef struct {
+   int hour;
+   int minute;
+} clock_t;
 
-void clock(time_text_t time_text, int hour, int minute);
-void clock_add(time_text_t time_text, int minute_offset);
+void clock_create(clock_text_t clock_text, clock_t clock);
+void clock_add(clock_text_t clock_text, int minute_offset);
+void clock_subtract(clock_text_t clock_text, int minute_offset);
+bool clock_is_equal(clock_t a, clock_t b);
```

Comment:
The API here still feels pretty unintuitive. Given that the `clock_text_t` type hides the fact that `clock_text` is actually an output parameter, I actually had to do a double-take to understand `clock_create`. Additionally, using `clock_text_t` instead of `clock_t` for the other methods feels a bit wonky.

How about something like this?

```c
clock_t clock_create(int hour, int min);
clock_text_t clock_to_string(clock_t clock); // I'm pretty sure this will just work
clock_t clock_add(clock_t clock, int minute_offset);
clock_t clock_subtract(clock_t clock, int minute_offset);
bool clock_is_equal(clock_t a, clock_t b);
```

I could go either way on using `clock_t` as a value or passing by reference so imagine that part however you'd prefer. This fully eliminates the use of the rendered state of a clock (a `clock_text_t`) as an awkward stand-in for a clock (ie: a `clock_t`).

Thoughts?

Reply:
With you on the output being more intuitive as the return value rather than an argument - but it does mean that the student will need to create the string from scratch. On the other hand it reduces the string handling to the one `clock_to_string()` method. I'll code it up and see what it looks like 😃

Reply:
I recall the reason that the strings are the output -> it is per the exercise specification.
If we stick to but retain the output value by return we end up with:

    clock_test_t clock_create(int hour, int minute);
    clock_text_t clock_add(clock_text_t clock, int minute offset);
    clock_text_t clock_subtract(clock_text_t clock, int minute offset);
    bool clock_is_equal(clock_text_t a, clock_text_t b);

The `clock_t` type disappears from the interface again.

Reply:
One option is to simply deviate a bit from the specifications. I think that we still meet the intent of the cases even if we deviate from the interface a bit. For example, see the sample solution for this problem in Lua: https://github.com/exercism/lua/blob/master/exercises/clock/example.lua.

However, if we want to match the upstream problem specification more closely, I like your last suggestion. It's awkward to have a string representation be the primary interchange type, but at least the tension between `clock_text_t` and `clock_t` has been resolved and it's that there _is_ a single dominant representation.

I'm happy with either approach.

Reply:
Neglected to think the string version completely through, it would require the student to manage memory allocation and freeing. As to currently an array is used as the input/output parameter. Changing to returning the string requires changing to pointers which introduces memory handling 😒

Reply:
You should can pass/return an array by value if you wrap it in a struct :+1:

Reply:
Of course 🙈